docs/source/notes/numpy_compatibility.rst

```diff
@@ -0,0 +1,388 @@
+
+   Since ``np.asarray()`` depends on the implementation of
+   ``torch.Tensor.__array__()`` which calls ``torch.numpy()``, the **leaf node**
+   requirement still needs to be satisfied by the user, i.e., ``requires_grad ==
+   False``
+
+From ``numpy``
+^^^^^^^^^^^^^^
+
+For a NumPy ``ndarray`` to be convertible to a PyTorch tensor:
+
+- It must have only native byte order
+- Must not be read only
+- Array strides must be multiples of the Torch element byte size
+- Must have a ``dtype`` which is one of ``float64 float32 float16 complex64
+  complex128 int64 int32 int16 int8 uint8 and bool``
+
+Concretely, these may be expressed as:
+
+.. doctest::
+
+   >>> a_np = np.ones(4).reshape(2, 2)
+   >>> b_torch = torch.tensor(a_np)
+   >>> assert(a_np.dtype.byteorder == '=') # Native byte order
+   True
+   >>> assert(a_np.flags.writeable == True) # Not read only
+   True
+   >>> np.equal([stride % b_torch.element_size() for stride in a_np.strides], np.zeros(len(a_np.strides))) # Multiples of torch element byte size
+   array([ True,  True])
+   >>> a_np.dtype in ["float64", "float32", "float16", "complex64", "complex128", "int64", "int32", "int16", "int8", "uint8", "bool"] # Supported dtype
+   True
+
+To obtain a **view** of the data, :meth:`torch.from_numpy()` can be used.
+
+.. doctest::
+
+   >>> a_np = np.array([1, 2, 3], dtype = np.float64)
+   >>> b_torch = torch.from_numpy(a_np)
+   >>> # b_torch = torch.as_tensor(a_np) # see note
+   >>> b_torch[2] = 23
+   >>> b_torch
+   tensor([ 1.,  2., 23.], dtype=torch.float64)
+   >>> a_np[0] = 22
+   >>> b_torch # view, changes with a_np
+   tensor([ 22.,  2., 23.], dtype=torch.float64)
+   >>> np.array_equal(b_torch.numpy(), a_np)
+   True
+
+- :meth:`torch.from_numpy()` is guaranteed to share memory with NumPy.
+- :meth:`torch.as_tensor()` will try to stay away from copy operations, it
+  also has the effect of sharing memory. However, ``torch.as_tensor()`` has
+  slightly higher overhead as it checks and accepts other iteratable objects as
+  well, e.g. ``list`` objects.
+- :meth:`torch.from_dlpack()` called with a NumPy array (``np.version.version >=
+  1.20``) as its argument will also generate a ``torch.Tensor`` view.
+
+To obtain a **copy** of the ``ndarray`` object, and not share memory, the
+:meth:`torch.tensor()` constructor accepts :meth:`np.ndarray` objects as a data
+source to construct and return a ``torch.Tensor``.
+
+.. note::
+
+   Recall that, if ``x`` is a tensor, ``torch.tensor(x)`` is equivalent to
+    ``x.clone().detach()``.
+
+.. doctest::
+
+   >>> a_np = np.array([1, 2, 3], dtype = np.float64)
+   >>> b_torch = torch.tensor(a_np)
+   >>> b_torch
+   tensor([1., 2., 3.], dtype=torch.float64)
+   >>> b_torch[2] = 23
+   >>> np.array_equal(b_torch.numpy(), a_np)
+   False
+   >>> c_np = b_torch.numpy().copy() # new memory
+
+Calling NumPy Functions on PyTorch
+----------------------------------
+
+Operators aside, NumPy functions can be called on PyTorch tensors as well. This
```

Comment:
This is not generally true. NumPy operations don't work on CUDA tensors, for example.

I'm not even sure this is true for CPU tensors. Don't some NumPy operations call methods on the original object and our implementation of those methods is (unfortunately) divergence from NumPy's?

Reply:
Yes, this is indeed the case for (IIRC) 6 reductions, like `max`/`mean`/`std`:
```python
>>> t = torch.tensor([1, 2, 3])
>>> np.std(t)
Traceback (most recent call last):
  File "<ipython-input-8-42c835b8b08a>", line 1, in <module>
    np.std(t)
  File "<__array_function__ internals>", line 180, in std
  File "/home/rgommers/anaconda3/envs/pytorch-111-nightly-jan2021/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 3613, in std
    return std(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)
TypeError: std() received an invalid combination of arguments 
```

Reply:
Those are the unfortunate exceptions, I think everything else does work.

Reply:
Rewording and adding exceptions